students/nam_vo/lesson04/mailroom_part2.py

```diff
@@ -0,0 +1,151 @@
+    ------------------------------------------------------------------
+    William Gates, III         $  653784.49           2  $   326892.24
+    Mark Zuckerberg            $   16396.10           3  $     5465.37
+    Jeff Bezos                 $     877.33           1  $      877.33
+    Paul Allen                 $     708.42           3  $      236.14
+
+    """
+    # Sort the donor's dictionary by total historical donation amount
+    sorted_donor_list = sorted(donor_dict.items(), key=sort_key, reverse=True)
+    # Print header
+    fheader = "{:<26}|{:^13}|{:^11}|{:>13}"
+    print(fheader.format("Donor Name", "Total Given", "Num Gifts", "Average Gift"))
+    print('-' * 66)
+    # Get body format
+    fbody = "{:<27}${:>11.2f}{:>12}  ${:>12.2f}"
+    # Loop thru each donor in the dictionary
+    for donor in sorted_donor_list:
+        # Get total donated amount, number of gifts, and average gift amount
+        total_given = sum(donor[1])
+        num_gifts = len(donor[1])
+        ave_gift = total_given / num_gifts
+        # Print report
+        print(fbody.format(donor[0], total_given, num_gifts, ave_gift))
+
+def send_all():
+    """Generate a thank you letter for each donor in the list and write each letter to current working directory as text file"""
+    # Loop thru each donor's name in the dictionary
+    for (name, donations) in donor_dict.items():
+        # Get the latest amount of donation
+        amount = donations[len(donations) - 1]
+        # Assign file name
+        fileName = name.replace(' ', '_').replace(',', '') + ".txt"
+        # Open file for writing
+        with open(fileName, 'w') as outFile:
+            # Construct letter's content
+            data = "{header}\n\n{body}\n\n{end}\n".format(**letter_template).format(name, amount)
+            # Write content to file
+            outFile.write(data)
+            # Print success message
+            print(f'File "{fileName}" has been created in current working directory!')
+
+def exit_program():
+    print("Bye!")
+    sys.exit()  # exit the interactive script
+
+def get_user_select(user_select):
+    """
+    Return a string contains user's selections
+
+    :user_select: dictionary of all user's selections
+    """
+    prompt = "\nChoose an action:\n\n"
+    for (key, val) in user_select.items():
+        prompt += f"{key} - {val}\n"
+    prompt += ">>> "
+
+    return prompt
+
+def sort_key(item):
+    """
+    Return total donated amount for selected donor, which will be used as key for the sorted() function.
+
+    :item: pair of donor's name and list of amounts they have donated
+    
+    """
+    return sum(item[1])
+   
+def print_donor(donor_dict):
+    """
+    Print all donor names.
+
+    :donor_dict: dictionary of donor's full name and list of amounts they have donated
+
+    """
+    # print("\n".join(["{}"]*len(donor_dict))).format(**donor_dict)
+    [print(name) for name in donor_dict]
```

Comment:
this is considered an anti-pattern. print() doesn't return a value so here are you creating a list with bunch of None values.

instead:
`print("\n".join(list(donor_dict)))` will give you the same result.